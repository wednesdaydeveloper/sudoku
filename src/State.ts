import * as sudoku  from './lib/SudokuSolver'

export interface SudokuCell extends sudoku.Sudoku.Cell {
  val: number;
  initial?: boolean;
};

export interface SudokuState {
  initialCells: SudokuCell[];
  resultCells: SudokuCell[];
  selectedCell: SudokuCell | undefined;
  hint: boolean;
  teacher: boolean;
};