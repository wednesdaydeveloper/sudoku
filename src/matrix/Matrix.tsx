import * as React from 'react';
import { SudokuCell } from '../State';
import './Matrix.css';

export interface StateProps {
  data: SudokuCell[];
  selected: SudokuCell | undefined;
}

export interface DispatchProps {
  onSelectCell: (cell: SudokuCell) => void;
}

export type Props = StateProps & DispatchProps;

const Matrix = (props: Props) => {
  const array: SudokuCell[][] = new Array<Array<SudokuCell>>(9);
  for (let row = 0; row < 9; row++) {
    array[row] = new Array<SudokuCell>(9);
    for (let col = 0; col < 9; col++) {
      const cell = props.data.find(c => c.col === col && c.row === row);
      array[row][col] = cell ? cell : { col, row, val: 0 };
    }
  }

  const isSelected = (cell: SudokuCell) => {
    return props.selected && cell.row === props.selected.row && cell.col === props.selected.col;
  };

  const _click = (e: React.MouseEvent<HTMLTableDataCellElement>, cell: SudokuCell) => {
//    e.preventDefault();
    if (!cell.initial) {
      props.onSelectCell(cell);
    }
  };

  const getClassName = (cell: SudokuCell): string => {
    let className = '';
    if (isSelected(cell)) {
      className = className + 'selected ';
    }
    if (cell.initial) {
      className = className + 'initial ';
    }
    return className;
  };
  return (
    <div>
      <table className="sudokuTable" >
        <tbody>
          {array.map((row, i) => (
            <tr key={i}>
              {row.map((cell, j) => (
                <td key={j} onClick={e => _click(e, cell)} className={getClassName(cell)}>
                  <p >{cell.val < 1 ? '' : cell.val}</p>
                </td>))}
            </tr>))}
        </tbody>
      </table>
    </div>
  );
};
export default Matrix;