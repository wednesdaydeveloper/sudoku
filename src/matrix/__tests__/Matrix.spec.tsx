import * as React from 'react';
import {shallow} from 'enzyme';

import Matrix, {Props} from '../Matrix';
import { SudokuCell } from '../../State';

describe('Matrix', () => {

  let props: Props;

  beforeEach(() => {

    const data: SudokuCell[] = Array<SudokuCell>(81);
    for (let row=0; row<9; row++) {
      for (let col=0; col<9; col++) {
        const val = row * 9 + col + 1;
        data[val-1] = {row, col, initial: true, val};
      }
    }
    //  デフォルトのプロパティ
    props = {
      data,
      selected: undefined ,
      onSelectCell: (cell: SudokuCell) => {return; },
    };
    spyOn(props, 'onSelectCell');
    
  });

  it('rendering', () => {
    const wrapper = shallow(<Matrix {...props} />);
    expect(wrapper.find('table').length).toBe(1);
    expect(wrapper.find('table').at(0).prop('className')).toBe('sudokuTable');

    expect(wrapper.find('tr').length).toBe(9);
    expect(wrapper.find('tr > td').length).toBe(81);
    for (let i=0; i<81; i++) {
      expect(wrapper.find('tr > td').at(i).prop('className')).toBe('initial ');
    }
    expect(wrapper.find('tr > td > p').length).toBe(81);
    for (let i=0; i<81; i++) {
      expect(wrapper.find('tr > td > p').at(i).prop('children')).toBe(i + 1);
    }
  });

  it('rendering valが０の場合は、数値を表示しない。', () => {
    props.data[10].val = 0;

    const wrapper = shallow(<Matrix {...props} />);

    expect(wrapper.find('tr > td > p').at(10).prop('children')).toBe('');
  });


  it('rendering initialがfalseの場合', () => {
    props.data[10].initial = false;

    const wrapper = shallow(<Matrix {...props} />);

    expect(wrapper.find('tr > td').at(10).prop('className')).toBe('');
    expect(wrapper.find('tr > td > p').at(10).prop('children')).toBe(11);
  });


  it('rendering selected が設定されている', () => {
    props.selected = {row: 1, col: 1, val: 999};  //  val は関係ない

    const wrapper = shallow(<Matrix {...props} />);

    expect(wrapper.find('tr > td').at(10).prop('className')).toBe('selected initial ');
    expect(wrapper.find('tr > td > p').at(10).prop('children')).toBe(11);
  });


  it('rendering selected が設定されていて initial がfalse', () => {
    props.data[10].initial = false;
    props.selected = {row: 1, col: 1, val: 999};  //  val は関係ない

    const wrapper = shallow(<Matrix {...props} />);

    expect(wrapper.find('tr > td').at(10).prop('className')).toBe('selected ');
    expect(wrapper.find('tr > td > p').at(10).prop('children')).toBe(11);
  });

  it('rendering when clicked initial がfalse', () => {
    props.data[10].initial = false;

    const wrapper = shallow(<Matrix {...props} />);

    wrapper.find('tr > td').at(10).simulate('click');
    
    expect(props.onSelectCell).toHaveBeenCalledWith(props.data[10]);
  });


  it('rendering when clicked initial がtrue', () => {
    props.data[10].initial = true;

    const wrapper = shallow(<Matrix {...props} />);

    wrapper.find('tr > td').at(10).simulate('click');
    
    expect(props.onSelectCell).not.toHaveBeenCalled();
  });

});