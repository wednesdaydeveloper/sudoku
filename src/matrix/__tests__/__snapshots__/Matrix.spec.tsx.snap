// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matrix Matrix „ÅÆ Snapshot 1`] = `
<div>
  <table
    className="sudokuTable"
  >
    <tbody>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            1
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            2
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            3
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            4
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            5
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            6
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            7
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            8
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            9
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            10
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            11
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            12
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            13
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            14
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            15
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            16
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            17
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            18
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            19
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            20
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            21
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            22
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            23
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            24
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            25
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            26
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            27
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            28
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            29
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            30
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            31
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            32
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            33
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            34
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            35
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            36
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            37
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            38
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            39
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            40
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            41
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            42
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            43
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            44
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            45
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            46
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            47
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            48
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            49
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            50
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            51
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            52
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            53
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            54
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            55
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            56
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            57
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            58
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            59
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            60
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            61
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            62
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            63
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            64
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            65
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            66
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            67
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            68
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            69
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            70
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            71
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            72
          </p>
        </td>
      </tr>
      <tr>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            73
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            74
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            75
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            76
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            77
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            78
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            79
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            80
          </p>
        </td>
        <td
          className="initial "
          onClick={[Function]}
        >
          <p>
            81
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
